### PR TITLE
[master] Removing conversions for SOA tables in Finite Check

### PR DESCRIPTION
**Pull request by owerbat**
New versions of ```computeSum()``` and ```checkFiniteness()``` have been created for SOA tables

### CODE REVIEW

File: service/kernel/data_management/finiteness_checker.cpp [hunk 1]
```diff
@@ -316,60 +491,56 @@ services::Status allValuesAreFiniteImpl(NumericTable & table, bool allowNaN, boo
...
     const NTLayout layout  = table.getDataLayout();
 
-    size_t nDataPtrs, nElementsPerPtr;
     if (layout == NTLayout::soa)
     {
-        // SOA layout: pointer for each column
-        nDataPtrs       = nColumns;
-        nElementsPerPtr = nRows;
+        // first stage: compute sum of all values and check its finiteness
+        double sum       = computeSumSOA<cpu>(table);
```
Comment (Alexsandruss) [inline service/kernel/data_management/finiteness_checker.cpp:497]: There is no need to rewrite code significantly and add specific function.
My concept of conversions removal:
1) Add type check from NumericTableDictionaryPtr
2) If type is applicable, get block of rows or column values
3) Pass pointer with data to function that will process it

`size_t nDataPtrs, nElementsPerPtr;`
`if (layout == NTLayout::soa)`
`{`
`        // SOA layout: pointer for each column`
`        nDataPtrs       = nColumns;`
`        nElementsPerPtr = nRows;`
`}`
`else`
`{`
`        // AOS layout: one pointer for all data`
`        nDataPtrs       = 1;`
`        nElementsPerPtr = nRows * nColumns;`
`}`
`NumericTableDictionaryPtr tableFeaturesDict = table.getDictionarySharedPtr();`
`for (size_t i = 0; i < nDataPtrs; ++i)`
`    switch ((*tableFeaturesDict)[j].getIndexType())`
`     case daal::data_management::features::IndexNumType::DAAL_FLOAT32:`
`          table.getBlockOfColumnValues()`
`          ...`
`          sum = computeSum(dataPtr, ...)`

**Comment by Alexsandruss** [comment]
Please, rebase from mnt branch to get last updates

**Comment by owerbat** [comment]
@Alexsandruss I've created [PR](https://github.com/oneapi-src/oneDAL/pull/638) for mnt branch with last updates

File: cpp/daal/src/data_management/finiteness_checker.cpp [hunk 2]
```diff
@@ -80,6 +83,41 @@ DataType computeSum(size_t nDataPtrs, size_t nElementsPerPtr, const DataType **
...
+            const float * colPtr = colBlock.get();
+            sum += computeSum<float, cpu>(1, nRows, &colPtr);
+            break;
+        }
+        case daal::data_management::features::IndexNumType::DAAL_FLOAT64:
+        {
+            ReadColumns<double, cpu> colBlock(table, i, 0, nRows);
+            DAAL_CHECK_BLOCK_STATUS(colBlock);
+            const double * colPtr = colBlock.get();
+            sum += static_cast<double>(computeSum<double, cpu>(1, nRows, &colPtr));
```
Comment (Alexsandruss) [inline cpp/daal/src/data_management/finiteness_checker.cpp:111]: Why static_cast? computeSum<double... already returns double.

File: cpp/daal/src/data_management/finiteness_checker.cpp [hunk 3]
```diff
@@ -158,6 +230,68 @@ double computeSum<double, avx512>(size_t nDataPtrs, size_t nElementsPerPtr, cons
...
+    });
+    tlsSum.reduce([&](double * localSum) { sum += *localSum; });
+
+    if (breakFlag)
+    {
+        sumIsFinite = false;
+    }
+    else
+    {
+        tlsFinite.reduce([&](bool * localFinite) { sumIsFinite &= !*localFinite; });
```
Comment (Alexsandruss) [inline cpp/daal/src/data_management/finiteness_checker.cpp:283]: Make sure that threading works correctly by running tests and adding NaN value in any column of numpy ndarray.

File: cpp/daal/src/data_management/finiteness_checker.cpp [hunk 4]
```diff
@@ -158,6 +230,68 @@ double computeSum<double, avx512>(size_t nDataPtrs, size_t nElementsPerPtr, cons
...
+        {
+            ReadColumns<double, avx512> colBlock(table, i, 0, nRows);
+            DAAL_CHECK_BLOCK_STATUS_THR(colBlock);
+            const double * colPtr = colBlock.get();
+            *localSum += static_cast<double>(computeSum<double, avx512>(1, nRows, &colPtr));
+            break;
+        }
+        }
+
+        *localFinite |= valuesAreNotFinite(localSum, 1, false);
```
Comment (Alexsandruss) [inline cpp/daal/src/data_management/finiteness_checker.cpp:268]: Variable name `localFinite` is confusing while reading code. It looks like `local finiteness = local finiteness OR valuesAreNotFinite`. Any variable name like `localNotFinite` or `localNonFiniteness` is better.

File: cpp/daal/src/data_management/finiteness_checker.cpp [hunk 5]
```diff
@@ -295,6 +429,64 @@ bool checkFiniteness<double, avx512>(const size_t nElements, size_t nDataPtrs, s
...
+        bool * localFinite = tlsFinite.local();
+
+        switch ((*tableFeaturesDict)[i].getIndexType())
+        {
+        case daal::data_management::features::IndexNumType::DAAL_FLOAT32:
+        {
+            ReadColumns<float, avx512> colBlock(table, i, 0, nRows);
+            DAAL_CHECK_BLOCK_STATUS_THR(colBlock);
+            const float * colPtr = colBlock.get();
+            *localFinite |= !checkFiniteness<float, avx512>(nRows, 1, nRows, &colPtr, allowNaN);
```
Comment (Alexsandruss) [inline cpp/daal/src/data_management/finiteness_checker.cpp:452]: Same there.

File: cpp/daal/src/data_management/finiteness_checker.cpp [hunk 6]
```diff
@@ -307,58 +499,59 @@ services::Status allValuesAreFiniteImpl(NumericTable & table, bool allowNaN, boo
     const size_t nElements = nRows * nColumns;
     const NTLayout layout  = table.getDataLayout();
 
-    size_t nDataPtrs, nElementsPerPtr;
+    size_t nDataPtrs;
```
Comment (Alexsandruss) [inline cpp/daal/src/data_management/finiteness_checker.cpp:502]: Delete nDataPtrs and connected code if they aren't used elsewhere.

Comment (owerbat) [reply cpp/daal/src/data_management/finiteness_checker.cpp]: Done

File: cpp/daal/src/data_management/finiteness_checker.cpp (hunk 4 above)
Comment (owerbat) [reply cpp/daal/src/data_management/finiteness_checker.cpp]: Fixed

File: cpp/daal/src/data_management/finiteness_checker.cpp (hunk 5 above)
Comment (owerbat) [reply cpp/daal/src/data_management/finiteness_checker.cpp]: Fixed

File: cpp/daal/src/data_management/finiteness_checker.cpp (hunk 2 above)
Comment (owerbat) [reply cpp/daal/src/data_management/finiteness_checker.cpp]: Fixed

**Pull request merged**